enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+  missing-metadata-version? -. No .-> exit
+
+  do-data-migration --> data-migration?
+  data-migration? -- Yes --> run
+  data-migration? -. No  .-> exit
+
+  run --> healthy?
+  healthy? -- Yes --> green -->  continue-running
+  healthy? -- Yes --> do-persist-version
+  healthy? -. No  .-> red
+
+  exit -.-> red
+  red --> reboot
+```
+
+### API Extensions
+
+Metadata persisted on filesystem related to the functionality described 
+in this enhancement such as version metadata, next boot action, and any other,
+is considered internal implementation detail and not an API to be consumed.
+Content (schema) and location of these files are subject to change.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Being a GA feature from the beginning the risks are not foreseeing fail scenarios in advance and implementation bugs
+that are not caught and fixed through graduation process.
+
+To mitigate the risks, a thorough review of the enhancement must be done by MicroShift, OpenShift, and RHEL teams,
+and making sure testing strategy is sound and prioritized equally with the feature development.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+### Definitions
+
+- **ostree commit**: TODO
+- **ostree deployment**: TODO
+- **Rollback**: booting previous ostree commit due to greenboot assessing system to not be functional
+- **Upgrade**: running newer version of MicroShift than previously as a result of booting another ostree commit
+- **Downgrade**: running older version of MicroShift than previously as a result of booting another ostree commit, not as part of a Rollback
+- **Backup**: backing up `/var/lib/microshift`
+- **Restore**: restoring `/var/lib/microshift`
+- **Version Metadata**: File residing in MicroShift data dir containing version of MicroShift that successfully started (cluster was healthy)
+- **MicroShift greenboot healthcheck**: Program verifying the status of MicroShift's cluster
+
+### Preface
+
+Every action related to procedure described in this enhancement is 
+performed after system's boot rather than immediately before shutdown. 
+Greenboot's healthchecks, green and red scripts are executed independent of MicroShift's processes.
+Actions related to backup, restore, and data migration may be performed with MicroShift 
+partially running, i.e. only etcd and kube-apiserver are running.
+
+Only one backup of MicroShift data will be stored at a given moment
+due to high probability of devices having limited storage.
+
+### Integration with greenboot
+
+Depending on result of greenboot's healthcheck either "green" (successful boot) or "red" (unsuccessful) scripts are executed before rebooting the system.
+
+MicroShift will integrate with that system to persist an action to perform on next boot:
+- "green": on next boot, before MicroShift starts, make a backup of MicroShift's data
+- "red": on next boot, before MicroShift starts, restore MicroShift's data from backup
```

Comment:
"red" is run every time the greenboot check fails. We should only set the "restore" flag if:
- There is a backup to restore from (delete /var/lib/microshift otherwise, i.e. first boot failed?)
- We are in the end of the reboot cycle, i.e. boot_counter=0

Reply:
Can you provide reasoning why should we do it?

Reply:
We can go with your current suggestion and see where it leads us. Let's just spell it out that there is a "red" loop and we're starting with a "clean slate" for the DB on each iteration.